### PR TITLE
Initial interface for business logic

### PR DESCRIPTION
**Pull request by JeffBlodgett**
Took longer than I expected to get familiar with IntelliJ IDEA today.  I've named the project "FileFilter" but we could revise that at any time.  I've included two interface files for review.  "IFileEnumerator" is centered about enumerating (getting a list) of files in a given folder and generating a list of duplicates found in the source.  I'm thinking that a Comparator defining how we want to test for file similarity could be passed into a FileEnumerator instance for use by the method that identifies duplicate files.

I'll start working on the proposal this afternoon, if that is ok, since I have some statements to make about the "problems being solved".  You could each edit it further after I check it in tomorrow morning (I'll do so by 9:00 AM).  Assuming you agree to the interfaces thus far, I can flesh out some stub classes and tests for the two interfaces above - the goal being to get things compiling and tests passing as quickly as possible so one of us can begin coding the GUI (we'll need to further discuss what the GUI will look like soon - perhaps exchange some hand drawings).

At the moment I'm at a bit of a loss in what could/should be serialized.  As you'll see in my initial take on the proposal, I foresaw this tool as (a) able to identify duplicate files in a given folder and (b) being able to copy that folder with or without particular duplicates to a given target path.  What is useful about these things?  Well, writing this utility in Java allows us to overcome the arbitrary 256-character pathname limit inherent in Windows.

I'm not sure what about points (a) and (b) would be useful if serialized and reloaded later.  We may have to add additional functionality, such as the ability to compare a previous serialized file list with a freshly generated list.  Then we could be talking about reporting on those files that have changed (new, modified, or missing/deleted).  This would be a different mode of operation- a mode "b".

### CODE REVIEW

**Comment by programming2web** [comment]
Hey Jeff, 
Probably it would be nice if the program has an option to compare files in 2 different filepaths and report the duplicates rather than only in 1 filepath. It would be useful for fast backups since only the files that have changed will need to be overriden. Somehow it's hard for me to imagine the situation why I would want to find duplicates in the same filepath but it could be an option as well. 
For finding duplicates in 2 filepaths we'll probably need a Diff class which could accept 2 filepaths and options. I think it could call Enumerator class to get the list of files in each of the filepaths and first compare the number of files and filenames in provided filepaths. Then if it finds the same filename in both filepaths it could compare if file revision / contents is the same. It might have several report methods like reportFileQtyDiff, reportFileVersionDiff. In GUI we could display it either by highlighting the diff files in both filepaths (can use color coding for various diff types: red = file is missing in other filepath, orange = file versions differ, blue = file contents is the same but filename is different, etc.). Or we could print a traditional report as text. Or even provide both options for user to choose from. 
If we have time to implement copying then we could provide many options to user so he could choose whether he wants to copy only non-duplicates or if filename differs but contents is the same - whether he wants to override it or not, etc. If we find duplicates within the source filepath we might ask user how he wants to handle that as well. 
Please let me know your thoughts. I could prepare the Diff class interface.
If you both don't mind I'd like to work more on GUI side of it. 
Thanks! Looking forward!

**Comment by dbendig** [comment]
I agree with Olga that there should be a way to provide both a target and source path. The GUI should take care of defaulting to the target and source being the same path, and the interfaces can even have methods that only take one path as a parameter, but with very little work involved we have a much more functional application with that separation.

I think we need to better qualify what type of duplicates we're looking for. If we make a utility that relies on file name and metadata and has copying functionality then we are really making a directory sync, rather than a duplicate check. A duplicate check should generate some sort of signature that can be compared across files to check that their contents are the same. The file system enforces against duplicate file names in one directory, although if our goal was to find the exact file within two directories, or within a single directory tree, then a simple file name check is a good place to start. Either way, we should pick a limited scope for the project so it should be a duplicate check or a directory sync, not both (at least not for this sprint.)

**Comment by JeffBlodgett** [comment]
Posted a "Proposal.docx" (MS Word 2010 format) draft proposal for you to review.  Please review and comment here with ideas (additions, downvotes/upvotes for certain features, etc.).

Also made some other changes in this "LogicInterface" branch this morning, as follows:
- Changed project name from "FileFilter" to "FileSieve".  "File Sieve" seems to be more abstract in meaning and sounds better (to me).
- Restructured project folders a little (per guidelines) and created a couple packages to hold the interfaces made thus far.
- Project now compiles successfully with ANT.  (This was critical - I had some odd errors yesterday)

I'll add methed(s) to the IFileManager interface for comparing two folders shortly, per this morning's Google Hangouts discussions.

I have a visual idea (GUI related) for how the concept of folder "sync" can be combined with finding duplicates in a single (source) folder.  I'll post a drawing once Olga starts a new branch with regards to the GUI (I'd suggest a single folder at the project root into which we can place and share drawings).

**Comment by smarks** [comment]
Nice work everyone so far. This is precisely what should be happening at this phase of project.

**Comment by JeffBlodgett** [comment]
As discussed on Google Hangouts today, I made some major refinements to the interfaces this morning.  Please review (they are well commented) and feel free to merge this "LogicInterfaces" branch. That way we can divvy up some work and checkout new branches for creating stubs, initial tests, and start to flesh out some classes.

Made a small change to the build.xml file (includeantruntime="false" in compile target) to eliminate a minor warning and added the /lib/ folder, with JUnit JAR and dependency, to Git for Ant's sake.  All Ant targets compile successfully (all, clean, compile, init, and test).  Gitignore file is such that all Netbeans or IDEA related files/folders are ignored by Git (I cannot speak for Eclipse).

File: src/main/java/FileSieve/BusinessLogic/FileManagement/IFileManager.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    /**
+     * Method for copying a file or empty folder to a target. Implementers should perform the copy operation
+     * on a separate (non-blocking) thread.
+     *
+     * @param sourcePathname    pathname of file or folder to copy
+     * @param targetPathname    pathname of file or folder to write
+     * @param overwriteFile     indicates if existing target should be overwritten (applies only to files)
+     */
+    public void copyFile(Path sourcePathname, Path targetPathname, boolean overwriteFile);
+
```
Comment (programming2web) [inline src/main/java/FileSieve/BusinessLogic/FileManagement/IFileManager.java:65]: public void deleteFile(Path pathname);

**Comment by programming2web** [comment]
Great work, Jeff!
Maybe it would be beneficial to also add deleteFile method to iFileManager.

**Comment by programming2web** [comment]
Also I would vote to make "Ability to compare files in a target folder with those in a reference folder and copy those that are missing or have changed from the reference folder to the target folder." an optional feature for now in the proposal.
Other than that it's all wonderful! Thanks for putting all this together!

**Comment by JeffBlodgett** [comment]
Good point about the addition of a delete file method Olga.  I'll add that plus an openFile method this evening.  Perhaps I can merge this branch later this evening after you both have time to comment further.  openFile would open the given file with whichever app is registered to open that file type.

**Comment by JeffBlodgett** [comment]
Just posted some more changes, mainly copyPathname (2), copyPathnames, deletePathname, and openPathname.  The copyPathnames and one of the copyPathname methods returns a single or list of SwingWorker instances.  Each SwingWorker can permit the copy operation of its given pathname (file or creation of a folder) to occur on a separate thread and it can provide progress messages to a PropertyChangeListener.  I provided a link with an excellent example in the javadoc comments.  The other copyPathname method is for simply copying on the callers thread... we can choose to implement some or all of these or none.

Let me know if I should merge this "LogicInterfaces" branch.  I feel it should be merged soon so we can say we accomplished week one goals (Git set-up, interfaces, and proposal) and be able to move on to stubs, tests, and some fleshing out of classes on several branches.

**Comment by JeffBlodgett** [comment]
Woke up this morning thinking it would be good for the copyPathnames method to return a Map rather than a list so the pathnames could be returned as keys (with SwingWorkers as values) to assist caller in tracking work.  Made this changes, and then renames the copyPathname and copyPathnames methods (those that return SwingWorker instances) to "getPathnameCopyWorker" and "getPathnameCopyWorkers".

Please review interfaces again when you two have a chance... I'll merge this "LogicInterfaces" branch into master when given the ok.  Won't be doing much more this weekend as I need to get caught up on the week 3 reading (also, I do most of my work during the week from the office, between support calls from the "peeps").  As previously mentioned (perhaps via Hangouts) I'd like your blessings to create an IFileDifferentiator stub class and test class (with tests) for it on a "FileDiff" branch.

One thing I will do this weekend it get to know IntelliJIDEA a little better.  I've been doing all Git adds, commits, pushes, and other actions from the "Git Bash" prompt (these days, I feel more comfortable at a Linux prompt than at a DOS prompt :-)).

**Comment by smarks** [comment]
This is a great interchange so far. NIce work. 

Interfaces should not begin with "I" that makes code harder to read. The fact that an interface is an interface should be easily discerned by its usage. The interface should a general concept that has specific implementations. The name should reflect this. It is also pretty easy to look at file while writing code to determine if it is an interface or not.  Lastly the JDK does not use "I" for Interfaces nor is this a convention listed in the Java Coding Conventions (which you should all be following). see http://www.oracle.com/technetwork/java/index-135089.html 

Question: Can anyone tell me the difference between an interface and abstract class? When and how are they used?

Keep up the good work - thanks.

**Comment by programming2web** [comment]
Jeff, it's great idea about the Map. I like it a lot!
Also thanks for sharing the link in regards to SwingWorker. This is very helpful. 
I'm totally fine with merging this pull request to master. It all looks great!

**Comment by JeffBlodgett** [comment]
Like a regular class, an abstract class can implement none, one, or all of an implemented interface's methods (if it implements an interface at all).  Like an interface, an instance of an abstract class cannot be instantiated so the class must be extended.  What makes the class abstract is one or more methods have been declared abstract and left without an implementation.  Other methods and data members implemented in the class may provide default implementations for certain functionality, but the class must be extended and implementations for all abstract methods provided in subclasses.

Per the professors comments, I have removed the "I" prefix from the interfaces.  I had learned in previous courses to use the I to clearly indicate a file defined an interface - I shall unlearn :-)

Merging this "LogicInterfaces" branch momentarily...

**Comment by JeffBlodgett** [comment]
Ok - that was silly... I failed to Git add, commit, and push the interface renames...

**Comment by JeffBlodgett** [comment]
I will open a "RenameInterface" branch for a moment, push the renames, and merge to handle this task...

**Comment by smarks** [comment]
Jeff, you did a good job describing the semantic differences between and abstract class and interface, But I am  also looking for how interfaces, abstract classes and sub classes are used - why do these semantic differences existing Java? What's their value? (Anyone besides Jeff care to have go at this?)

**Pull request merged**